### PR TITLE
FIX: read_dir and read_labspec with non readable files

### PR DESCRIPTION
**Pull request by atravert**
`read_dir()`: now ignores non-readable files (unknown extension, etc...) 
`read_labspec()`: now returns None if attempt is made to read a non-labspec `.txt` file (fairly common, especially when called from `read_dir()`

Check List for a good PR:
- [x] Tests added for labspec() and read_dir()  
- [x]  `CHANGELOG.md` updated

### CODE REVIEW

**Comment by fernandezc** [comment]
Ok I'll will make the review today. I just need to learn how to install your PR in a local branch on my computer to check it and add the test for None file in read_dir (this test is made in test_importer... so no need to add new test file: I just need to look if this case need additional test). 

***fs!*** this is a fake file system (see pyfakefs package) : the objective for the importer test is to avoid loading physical file that are absent when spectroscopy is installed using pypi.  Indeed, spectrochempy_data is a conda package and so it is not installed in this case). All test that requires loading experimental file should be preceded by something like : 

```python
from spectrochempy.core import preferences as prefs

DATADIR = prefs.datadir
IRDATA = DATADIR / "irdata"


@pytest.mark.skipif(
    not IRDATA.exists(),
    reason="Experimental data not available for testing",
)
def test_xxx():
    """
    Test using files in IRDATA
    """
```
I need to explain better how this work in dev guide, when I will have finished review our tests directory:  A lot of work which has already permitted to discover a lot of bugs in core files!   Work in progress....

**Comment by fernandezc** [comment]
Try on a local branch?  it's easy with GitHub CLI.

'''
conda install gh --channel conda-forge

gh auth login
gh pr checkout 363
'''

**Comment by fernandezc** [comment]
To test read_dir: 

Add this in test_importer.py 

Line 51
```python
elif filename.stem == "emptyfake":
      dataset = None
```
to simulate a reader that return `None`. 

Line 219
```python
f6 = DATADIR / "fakedir/emptyfake.fk"  # return None when reader
fs.create_file(f6)
```

Now you can see that Line 275
this still work:
```python
    nd = read_dir(DATADIR / "fakedir")
    assert nd.shape == (4, 3)
```
which means that the None are not added.

File: docs/devguide/codespec_adding_reader.rst [hunk 1]
```diff
@@ -9,7 +9,7 @@ Import of spectroscopic data with their meta data from various file formats is a
 import is made through the generic ``read()`` function (in ``core.readers.importer``) which handles
 the determination of filetype, search locations,  etc...).
 
-The steps to add a specific reader - with the example of Thermo Galactic .spc files - are as follows:
+In the following, we descrube the steps to add a specific reader - with the example of Grams/Thermo .spc files.
```
Comment (fernandezc) [inline docs/devguide/codespec_adding_reader.rst:12]: describe!

File: tests/test_core/test_readers/test_read_labspec.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 import pytest
 
 import spectrochempy as scp
-from spectrochempy.utils import show
+import os
```
Comment (fernandezc) [inline tests/test_core/test_readers/test_read_labspec.py:7]: for consistency I would avoid os and use pathlib
```python
from pathlib import Path
```

File: tests/test_core/test_readers/test_read_labspec.py [hunk 3]
```diff
@@ -15,8 +15,25 @@
...
+    # single file
     nd = scp.read_labspec("Activation.txt", directory=RAMANDIR)
     assert nd.shape == (532, 1024)
 
+    # with read_dir
     nd = scp.read_dir(directory=RAMANDIR / "subdir")
     assert nd.shape == (6, 1024)
+
+    # empty txt file
+    with open("i_am_empty.txt", "w") as f:
```
Comment (fernandezc) [inline tests/test_core/test_readers/test_read_labspec.py:27]:
```python
f = Path("i_am_empty.txt")
```

File: tests/test_core/test_readers/test_read_labspec.py [hunk 4]
```diff
@@ -15,8 +15,25 @@
...
+        f.close()
+    nd = scp.read_labspec("i_am_empty.txt")
+    os.remove("i_am_empty.txt")
```
Comment (fernandezc) [inline tests/test_core/test_readers/test_read_labspec.py:30]:
```python
f.unlink()
```

File: tests/test_core/test_readers/test_read_labspec.py [hunk 5]
```diff
@@ -15,8 +15,25 @@
...
+    assert nd is None
+
+    # non labspec txt file
+    with open("i_am_not_labspec.txt", "w") as f:
```
Comment (fernandezc) [inline tests/test_core/test_readers/test_read_labspec.py:34]:
```python
f = Path("i_am_not_labspec.txt")
f.write_text("blabla")
```

File: tests/test_core/test_readers/test_read_labspec.py [hunk 6]
```diff
@@ -15,8 +15,25 @@
...
+        f.write("blabla")
+        f.close()
+    nd = scp.read_labspec("i_am_not_labspec.txt")
+    os.remove("i_am_not_labspec.txt")
```
Comment (fernandezc) [inline tests/test_core/test_readers/test_read_labspec.py:38]: f.unlink()

File: CHANGELOG.md [hunk 7]
```diff
@@ -8,6 +8,10 @@
 
 * Issue #360 : write a dataset with a specified filename do not open a dialog except if the file already exists.
 
+* read_dir: now ignores non readable files
```
Comment (fernandezc) [inline CHANGELOG.md:11]: and read!  which also do directory reading (see example in test_importer)

File: spectrochempy/core/readers/importer.py [hunk 8]
```diff
@@ -492,13 +495,15 @@ def _read_dir(*args, **kwargs):
...
-    for key in files.keys():
+    valid_extensions = list(zip(*FILETYPES))[0] + list(zip(*ALIAS))[0]
+    for key in [key for key in files.keys() if key[1:] in valid_extensions]:
         if key:
             importer = Importer()
             nd = importer(files[key], **kwargs)
-            if not isinstance(nd, list):
-                nd = [nd]
-            datasets.extend(nd)
+            if nd is not None:
```
Comment (fernandezc) [inline spectrochempy/core/readers/importer.py:503]: Are you sure you need this. For me it works well without this?  Can you provide an example when it doesn't work?

File: CHANGELOG.md (hunk 7 above)
Comment (atravert) [reply CHANGELOG.md:11]: done

File: docs/devguide/codespec_adding_reader.rst (hunk 1 above)
Comment (atravert) [reply docs/devguide/codespec_adding_reader.rst:12]: done

**Comment by fernandezc** [comment]
Wait a minute! I did a mistake: my read_dir test is not correct!  I come back soon.

**Comment by fernandezc** [comment]
ok to have a correct test, in test_importer.py line 102, add 
```python
res.append(DATADIR / "fakedir/emptyfake.fk")
```
to simulate a complete reading of the directory. 

It seems to work like this without the additional check line (line 503): 
```if nd is not None:```
because the change in switch_protocol line 206 seems enough.

**Comment by atravert** [comment]
OK. I probably made the change first in 503 , and later in 206 :). Il will check again tonight !

Le 27 janvier 2022 15:00:34 GMT+01:00, Christian Fernandez ***@***.***> a écrit :
>ok to have a correct test, in test_importer.py line 102, add
>
>res.append(DATADIR / "fakedir/emptyfake.fk")
>
>to simulate a complete reading of the directory.
>
>It seems to work like this without the additional check line (line
>503):
>if nd is not None:
>because the change in switch_protocol line 206 seems enough.
>
>—
>Reply to this email directly, view it on
>GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fpull%2F363%23issuecomment-1023240049&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C6f48bb3b15db4c89087708d9e19d642a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637788888371152078%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=ZbHL4rqASSjf%2BIAZ6mkqpS%2FCOWYzwmtsZYTfCmU37Hg%3D&reserved=0>,
>or
>unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACNBNCQKADCWV6R5D6INSXDUYFFYFANCNFSM5M5K35DQ&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C6f48bb3b15db4c89087708d9e19d642a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637788888371152078%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=rh6dJNQAEeO82OVtR7Tg9jZAhmY3EnsEgwSEkVTiK%2BI%3D&reserved=0>.
>Triage notifications on the go with GitHub Mobile for
>iOS<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C6f48bb3b15db4c89087708d9e19d642a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637788888371152078%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=EIZWQBz9P6G6zAiBcG2gG2tIyVrt0xXj6vom55Tf%2BJA%3D&reserved=0>
>or
>Android<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7C6f48bb3b15db4c89087708d9e19d642a%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637788888371152078%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=VuOhtf6xpE92KYZSYZKSb7egwTri%2BYgMQxwOD8eyRvw%3D&reserved=0>.
>You are receiving this because you authored the thread.Message ID:
***@***.***>

-- Envoyé depuis /e/ Mail.

File: spectrochempy/core/readers/importer.py (hunk 8 above)
Comment (atravert) [reply spectrochempy/core/readers/importer.py:503]: Yes it was not necessary :-)
